### PR TITLE
[SQL] Support casts between VARIANT and ARRAY and MAP

### PR DESCRIPTION
**Pull request by mihaibudiu**
This is still not finished

### CODE REVIEW

**Comment by mihaibudiu** [comment]
The following generated program does not compile:
```
// Automatically-generated file
#![allow(dead_code)]
#![allow(non_snake_case)]
#![allow(unused_imports)]
#![allow(unused_parens)]
#![allow(unused_variables)]
#![allow(unused_mut)]
#![allow(unconditional_panic)]

#![allow(non_camel_case_types)]
#[cfg(test)]
use hashing::*;
use dbsp::{
    algebra::{ZSet, MulByRef, F32, F64, Semigroup, SemigroupValue, ZRingValue,
         UnimplementedSemigroup, DefaultSemigroup, HasZero, AddByRef, NegByRef,
         AddAssignByRef,
    },
    circuit::{checkpointer::Checkpoint, Circuit, CircuitConfig, Stream},
    operator::{
        Generator,
        FilterMap,
        Fold,
        group::WithCustomOrd,
        time_series::{RelRange, RelOffset, OrdPartitionedIndexedZSet},
        MaxSemigroup,
        MinSemigroup,
        CmpFunc,
    },
    OrdIndexedZSet, OrdZSet,
    TypedBox,
    utils::*,
    zset,
    indexed_zset,
    DBWeight,
    DBData,
    DBSPHandle,
    Error,
    Runtime,
    NumEntries,
    MapHandle, ZSetHandle, OutputHandle,
    dynamic::{DynData,DynDataTyped},
};
use feldera_types::program_schema::SqlIdentifier;
use dbsp_adapters::Catalog;
use feldera_types::{deserialize_table_record, serialize_table_record};
use size_of::*;
use ::serde::{Deserialize,Serialize};
use compare::{Compare, Extract};
use std::{
    collections::BTreeMap,
    convert::identity,
    ops::Neg,
    fmt::{Debug, Formatter, Result as FmtResult},
    path::Path,
    marker::PhantomData,
};
use core::cmp::Ordering;
use rust_decimal::Decimal;
use dbsp::declare_tuples;
use feldera_sqllib::{
    *,
    array::*,
    casts::*,
    binary::*,
    geopoint::*,
    timestamp::*,
    interval::*,
    string::*,
    operators::*,
    aggregates::*,
    variant::*,
};

#[cfg(not(target_env = "msvc"))]
#[global_allocator]
static ALLOC: tikv_jemallocator::Jemalloc = tikv_jemallocator::Jemalloc;

#[allow(non_upper_case_globals)]
#[export_name = "malloc_conf"]
pub static malloc_conf: &[u8] = b"prof:true,prof_active:true,lg_prof_sample:19\0";
#[cfg(test)]
use readers::*;

declare_tuples! {
}



pub fn circuit0(cconf: CircuitConfig) -> Result<(DBSPHandle, (OutputHandle<WSet<Tup1<Option<BTreeMap<String, Variant>>>>>, )), Error> {

    let (circuit, streams) = Runtime::init_circuit(cconf, |circuit| {
        // rel#0:LogicalValues.(type=RecordType(INTEGER ZERO),tuples=[{ 0 }])
        let stream36 = circuit.add_source(Generator::new(|| if Runtime::worker_index() == 0 {zset!(
            Tup1::new(0i32) => 1,
        )} else {zset!(
        )}));
        // rel#25:LogicalProject.(input=LogicalValues#0,exprs=[CAST(PARSE_JSON('{"a": 1}')):(VARCHAR NOT NULL, VARIANT NOT NULL) MAP])
        // DBSPMapOperator 65
        let stream65: Stream<_, WSet<Tup1<Option<BTreeMap<String, Variant>>>>> = stream36.map(move |t_1: &Tup1<i32>, | -> 
        Tup1<Option<BTreeMap<String, Variant>>> {
            Tup1::new(cast_to_mapN_V(
                parse_json_s(
                    String::from("{\"a\": 1}"))))
        });
        // CREATE VIEW `V` AS
        // SELECT CAST(PARSE_JSON('{"a": 1}') AS MAP< VARCHAR, VARIANT >)
        // DBSPSinkOperator 92(81)
        let handle92 = stream65.output();

        Ok((handle92, ))
    })?;
    Ok((circuit, streams))
}


#[test]
pub fn test0() {
    // VariantTests#testCastMap

    let mut circuitAndStreams = circuit0(
        CircuitConfig::with_workers(
            2usize)).unwrap(
    );
    let streams = circuitAndStreams.1;
    pub fn input0() -> () {
        ()
    }
    let input0: () = input0(
    );
    let _ = circuitAndStreams.0.step(
    ).unwrap(
    );
    assert!(
        must_equal(
            &read_output_handle(
                &streams.0),
            &zset!(
                Tup1::new(None::<BTreeMap<String, Variant>>) => 1,
            )),
        r#"
        mvn test -Dtest=VariantTests#testCastMap
        VariantTests#testCastMap"#);
}
```

**Comment by mihaibudiu** [comment]
Fixes #2498

**Comment by mihaibudiu** [comment]
@ryzhyk please review at least the Rust changes, they are not big
you can also review the last file, with the new tests, to see what the increased coverage is

File: crates/sqllib/src/casts.rs [hunk 1]
```diff
@@ -1591,18 +1592,9 @@ where
...
-    K: Clone,
-    V: Clone,
-{
-    map.into()
-}
-
 pub fn cast_to_vec_V<T>(value: Variant) -> Vec<T>
 where
-    Vec<T>: TryFrom<Variant>,
+    Vec<T>: TryFrom<Variant, Error = Box<dyn Error>>,
```
Comment (ryzhyk) [inline crates/sqllib/src/casts.rs:1597]: Would `<Vec<T>>::Error: Display` be sufficient at this point?

You might have to write it as
```
<Vec<T> as TryFrom<Variant>>::Error: Display,
```

**Review by ryzhyk — APPROVED** [review]

**Pull request merged**